include/astarte_device_sdk/formatter.hpp

```diff
@@ -0,0 +1,436 @@
+    return ctx.begin();
+  }
+
+  /**
+   * @brief Formats the AstarteDatastreamObject object as a key-value map.
+   * @param data The AstarteDatastreamObject to format.
+   * @param ctx The format context.
+   * @return An iterator to the end of the output.
+   */
+  template <typename FormatContext>
+  auto format(const AstarteDeviceSdk::AstarteDatastreamObject &data, FormatContext &ctx) const {
+    auto out = ctx.out();
+    out = NS_FORMAT::format_to(out, "{{");
+
+    bool first = true;
+    for (const auto &pair : data.get_raw_data()) {
+      if (!first) {
+        out = NS_FORMAT::format_to(out, ", ");
+      }
+      out = NS_FORMAT::format_to(out, "\"{}\": {}", pair.first, pair.second);
+      first = false;
+    }
+
+    out = NS_FORMAT::format_to(out, "}}");
+    return out;
+  }
+};
+
+/**
+ * @brief NS_FORMAT::formatter specialization for AstarteDeviceSdk::AstartePropertyIndividual.
+ */
+template <>
+struct NS_FORMAT::formatter<AstarteDeviceSdk::AstartePropertyIndividual> {
+  /**
+   * @brief Parses the format string. Default implementation.
+   * @param ctx The parse context.
+   * @return An iterator to the end of the parsed range.
+   */
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext &ctx) const {
+    return ctx.begin();
+  }
+
+  /**
+   * @brief Formats the AstartePropertyIndividual object.
+   * @param data The AstartePropertyIndividual to format.
+   * @param ctx The format context.
+   * @return An iterator to the end of the output.
+   */
+  template <typename FormatContext>
+  auto format(const AstarteDeviceSdk::AstartePropertyIndividual &data, FormatContext &ctx) const {
+    if (data.get_value().has_value()) {
+      return NS_FORMAT::format_to(ctx.out(), "{}", data.get_value().value());
+    }
+
+    return ctx.out();
+  }
+};
+
+/**
+ * @brief NS_FORMAT::formatter specialization for AstarteDeviceSdk::AstarteMessage.
+ */
+template <>
+struct NS_FORMAT::formatter<AstarteDeviceSdk::AstarteMessage> {
+  /**
+   * @brief Parses the format string. Default implementation.
+   * @param ctx The parse context.
+   * @return An iterator to the end of the parsed range.
+   */
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext &ctx) const {
+    return ctx.begin();
+  }
+
+  /**
+   * @brief Formats the AstarteMessage object.
+   * @param msg The AstarteMessage to format.
+   * @param ctx The format context.
+   * @return An iterator to the end of the output.
+   */
+  template <typename FormatContext>
+  auto format(const AstarteDeviceSdk::AstarteMessage &msg, FormatContext &ctx) const {
+    auto out = ctx.out();
+
+    out =
+        NS_FORMAT::format_to(out, "{{interface: {}, path: {}", msg.get_interface(), msg.get_path());
+
+    // check if the payload is an unset property, which is the only "empty" case
+    bool is_unset_prop = false;
+    if (const auto *prop =
+            std::get_if<AstarteDeviceSdk::AstartePropertyIndividual>(&msg.get_raw_data())) {
```

Comment:
As a general rule I would avoid assignments in if guards. They are often hard to read.